core/src/main/java/com/graphhopper/routing/weighting/FlexModelWeighting.java

```diff
@@ -0,0 +1,151 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.weighting;
+
+import com.graphhopper.routing.profiles.BooleanEncodedValue;
+import com.graphhopper.routing.profiles.EncodedValueFactory;
+import com.graphhopper.routing.profiles.EncodedValueLookup;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.routing.util.FlexModel;
+import com.graphhopper.routing.util.HintsMap;
+import com.graphhopper.routing.weighting.flex.AverageSpeedFlexConfig;
+import com.graphhopper.routing.weighting.flex.DelayFlexConfig;
+import com.graphhopper.routing.weighting.flex.PriorityFlexConfig;
+import com.graphhopper.util.EdgeIteratorState;
+
+import static com.graphhopper.util.EdgeIterator.NO_EDGE;
+
+/**
+ * Every EncodedValue like road_environment can influence one or more aspects of this Weighting: the delay, the
+ * speed_factor, the average_speed and the priority. The formula is basically:
+ * <pre>
+ * if no access to edge then return infinity
+ * seconds = toSeconds(distance / speed) + delayInSeconds
+ * distanceInfluence = distance * distanceFactor
+ * weight = (seconds + distanceInfluence) / priority;
+ * return weight
```

Comment:
What about elevation in this formula, is it not included for a reason ?

Reply:
We could add it here and make it a first class citizen. Another way could be to create two new EncodedValues up_elevation and down_elevation that would avoid fetching the pillar node information and directly be usable like the other properties (like max_width)

Reply:
Trying to get this straight on a bit more abstract level: So far the weight is composed as 

```weight
weight = a+b*distance
```
We have a constant term and one proportional to the distance. a and b are just composed by the other (per-edge) constants (`delayInSeconds,distanceFactor,priority,speed`). I am even wondering if all that `FlexModel` should provide is `a,b,a_min,b_min` (and maybe a few more, see below). Of course, since speed, distanceFactor, priority etc. are useful concepts it makes sense to have the default implementation of `FlexModel` take these as input (and calculate a,b from them). But maybe for another use-case one wants to use other factors, which would be easy to do by providing another `FlexModel` implementation to obtain a&b. 

Regarding the elevation of course this only moves the same question to `FlexModel` (or the default implementation of it). Storing up/down_elevation per edge as encoded values just like the others sounds like a good idea!

Btw I was briefly thinking about whether or not there should even be a quadratic term for the distance (c*distance^2). Its like asking for something like "do not use (or enforce) roads with long distances between junctions", which might be useful. Less junctions means safer driving, I guess.

Also a contribution `weight+=c/distance` could be useful: For example an edge might have up_elevation=100,down_elevation=-30 and we want to avoid routes that are steep. Then we could do `weight += some_factor*(up_elevation-down_elevation)/distance. Or in case we want to prefer routes that go up and down a lot we could add `weight += -some_factor*up_elevation/distance` (here c is negative).
Another complication here: The resulting weight must not be negative (the constants must not be ill-chosen for the possible range of distances). For the inverse contribution even more than for the distance it seems like a quadratic term could be useful, something like `weight += d/(distance^2)` and d could again be `up_elevation` to avoid very steep sections.

The most general formula would then be:
```
weight=a+b*distance+c*distance^2+d/distance+e/(distance^2)
```

With this in mind it seems like its a good idea to (only) have this formula in `FlexWeighting` and move the more specific stuff (priority, slope etc.) to `FlexModel` ?

Reply:
I forgot to take into account the difference between weight and time in this discussion. For example priority can affect the weight but not the time.

Reply:
Thanks for thinking this through :)

> But maybe for another use-case one wants to use other factors, which would be easy to do by providing another FlexModel implementation to obtain a&b.

At this point I would avoid any abstraction and just trying to make it working. E.g. performance is making problems especially with more than 1 entry. And also it is not clear for me if we can model most real world usages.

Additionally I would like to keep FlexModel very stupid but surely we can move stuff out of the Weighting into some WeightingModelController :)

> Less junctions means safer driving, I guess.

Interesting. But should it be more about turns and this is controlled via turn costs?

> Also a contribution weight+=c/distance could be useful

Ah, sure. In combination with elevation we really need this. Do we need this for other things? (If not, we could store it in the EncodedValues directly)

> (the constants must not be ill-chosen for the possible range of distances

This is a general problem: range checking. I think there are at least 3 types: one is avoiding negative to make dijkstra working, the other is ensuring correctness of A* (getMinWeight) and the last one is to ensure correctness for landmarks (do not allow weight decrease).

> I forgot to take into account the difference between weight and time in this discussion. For example priority can affect the weight but not the time.

Ah, yeah. This is another complexity here.

Reply:
> At this point I would avoid any abstraction and just trying to make it working. E.g. performance is making problems especially with more than 1 entry. And also it is not clear for me if we can model most real world usages.

Ok I understand you want to keep the configuration options as simple as possible to get started and deal with the issues arising from having any kind of flexible weighting first.

> Additionally I would like to keep FlexModel very stupid but surely we can move stuff out of the Weighting into some WeightingModelController :

Right now FlexModel is a pure data class which makes sense. I was thinking about 'converting' the 'user input' (this is represented by FlexModel atm) into the formula used by weighting internally. If we ever use the full formula as I wrote it will become messy to have all the priorities,slopes,speeds etc. right there in the formula and it would be nicer to first have something that calculates a,b,c,d from the things the user can choose from and pass these to the flexible weighting. This would also allow to do consistency checks for a,b,c,d (weight must not become negative).

> Interesting. But should it be more about turns and this is controlled via turn costs?

Avoiding junctions can be modeled using turn costs as well (adding costs for going straight), but then its not usable with node-based routing. But not sure how well it would work to use the quadratic distance term.

> Ah, sure. In combination with elevation we really need this. Do we need this for other things? (If not, we could store it in the EncodedValues directly)

You mean storing the 'up/down_slope` instead of just the 'up/down_elevation' ? Does it make a difference ? If we have the slope and distance we can calculate the elevation and if we have the distance and elevation we can calculate the slope.

Reply:
> If we ever use the full formula as I wrote it will become messy to have all the priorities,slopes,speeds etc. right there in the formula and it would be nicer to first have something that calculates a,b,c,d from the things the user can choose from and pass these to the flexible weighting. This would also allow to do consistency checks for a,b,c,d (weight must not become negative).

What do you mean here or can you show this as code? I'm not sure I understand the difference and why it would be messy.

> Does it make a difference ?

Not really but we would not need the additional distance^-1 term.